### PR TITLE
select() not creating duplicate variable on grouped data frames.

### LINKED ISSUE #5841: Renaming with select() creates duplicate variables with a grouped dataframe
Renaming a variable, in a select statement, with the same name as the grouping variable results in a dataframe with 2 variables with the same name. There is a warning that the grouping variable was added but no error about duplicated variable names until the variable is referenced.  

Not sure if this is the expected behavior, but it seemed weird to me. Especially since mutating a new variable with the same name as the existing grouping variable successfully replaces the original grouping variable.

``` r
library(tidyverse)
#> Warning: package 'tibble' was built under R version 4.0.4

(df <- tribble(
  ~prov,            ~value, ~prov_fr,
  "Ontario",         10,    "Ontario",
  "Quebec",          20,    "Québec",
  "Rest of Canada",  30,    "Reste du Canada"
) %>%
    group_by(prov) %>% 
    select(prov = prov_fr, value))
#> Adding missing grouping variables: `prov`
#> # A tibble: 3 x 3
#> # Groups:   prov [3]
#>   prov           prov            value
#>   <chr>          <chr>           <dbl>
#> 1 Ontario        Ontario            10
#> 2 Quebec         Québec             20
#> 3 Rest of Canada Reste du Canada    30

df %>% select(prov)
#> Error: Names must be unique.
#> x These names are duplicated:
#>   * "prov" at locations 1 and 2.


tribble(
  ~prov,            ~value, ~prov_fr,
  "Ontario",         10,    "Ontario",
  "Quebec",          20,    "Québec",
  "Rest of Canada",  30,    "Reste du Canada"
) %>%
  group_by(prov) %>% 
  mutate(prov = "a")
#> # A tibble: 3 x 3
#> # Groups:   prov [1]
#>   prov  value prov_fr        
#>   <chr> <dbl> <chr>          
#> 1 a        10 Ontario        
#> 2 a        20 Québec         
#> 3 a        30 Reste du Canada
```

<sup>Created on 2021-04-07 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### PR DESCRIPTION
**Pull request by romainfrancois**
closes #5841

### CODE REVIEW

File: tests/testthat/test-select.r [hunk 1]
```diff
@@ -140,6 +140,19 @@ test_that("select() keeps attributes of raw data frames (#5831)", {
   expect_equal(attr(select(df, x), "a"), "b")
 })
 
+test_that("select() not creating duplicate variables with grouped data frame (#5841)", {
```
Comment (hadley) [inline tests/testthat/test-select.r:143]: I'd suggest combining this with whatever test checks that variables are automatically included, and using `expect_snapshot()` to capture the message.

File: NEWS.md [hunk 2]
```diff
@@ -1,6 +1,9 @@
 # dplyr (development version)
 
+* `select()` not creating duplicate variable with a grouped data frame (#5841).
```
Comment (hadley) [inline NEWS.md:3]:
```suggestion
* `select()` no longer creates duplicate variables when renaming a variable to the same name as a grouping variable (#5841).
```

File: R/select.R [hunk 3]
```diff
@@ -137,13 +137,20 @@ ensure_group_vars <- function(loc, data, notify = TRUE) {
 
   if (length(missing) > 0) {
     vars <- names(data)[missing]
-    if (notify) {
+    added_group_loc <- set_names(missing, vars)
+
+    # not adding group variables that have the same name
+    # as something that is selected. #5841
```
Comment (hadley) [inline R/select.R:143]:
```suggestion
    # don't add grouping variables with same name as new variable (#5841)
```

**Review by hadley — COMMENTED** [review]
My only (minor) worry is that this changes the grouping, but that seems to be what the user is requesting with this code.

**Comment by romainfrancois** [comment]
🤔 maybe this should do what `mutate()` does and regroup based on the names: 

``` r
library(tidyverse, warn.conflicts = FALSE)

(df <- tribble(
  ~prov,            ~value, ~prov_fr,
  "Ontario",         10,    "Ontario",
  "Quebec",          20,    "Québec",
  "Rest of Canada",  30,    "Reste du Canada"
) %>%
    group_by(prov) %>% 
    select(prov = prov_fr, value))
#> # A tibble: 3 x 2
#>   prov            value
#>   <chr>           <dbl>
#> 1 Ontario            10
#> 2 Québec             20
#> 3 Reste du Canada    30

tribble(
  ~prov,            ~value, ~prov_fr,
  "Ontario",         10,    "Ontario",
  "Quebec",          20,    "Québec",
  "Rest of Canada",  30,    "Reste du Canada"
) %>%
  group_by(prov) %>% 
  mutate(prov = prov_fr)
#> # A tibble: 3 x 3
#> # Groups:   prov [3]
#>   prov            value prov_fr        
#>   <chr>           <dbl> <chr>          
#> 1 Ontario            10 Ontario        
#> 2 Québec             20 Québec         
#> 3 Reste du Canada    30 Reste du Canada
```

<sup>Created on 2021-04-09 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by hadley** [comment]
Yeah, that would mean we're at least following an existing behaviour rather than introducing a new one. OTOH this isn't exactly modifying an existing grouping variable; it's changing the name of a non-grouping variable.

**Comment by romainfrancois** [comment]
We could inform about changing the grouping because ...

Or refuse to rename a non grouping variable with the name of a grouping variable ?

now we are: Silently dropping a grouping variable that would otherwise have been added with a message"

**Pull request merged**